### ISSUE TITLE
Regex: use pcrejit

### ISSUE THREAD

**Issue by asterite** [crystal-lang/crystal#3852]
Use [pcrejit](http://www.pcre.org/original/doc/html/pcrejit.html) in Regex to speed up things.

I'm having a hard time making it work on Mac OSX because an old version of pcre comes installed and both clang and crystal pick up that library version (though it seems to be picking up the headers from hoembrew). For example:

```c
#include <pcre.h>
#include <stdio.h>

int main() {
  int where = 0;
  int ret = 0;
  ret = pcre_config(PCRE_CONFIG_JIT, &where);
  printf("%d\n", ret);
  printf("%d\n", where);
  return 0;
}
```

Compiling and running it:

```
$ clang c.c -lpcre
$ ./a.out 
-3
0
# The above means JIT is not supported

$ clang c.c -lpcre -L/usr/local/Cellar/pcre/8.39/lib
$ ./a.out 
0
1
# The above means JIT is supported
```

We obviously can't hardcode `/usr/local/Cellar/pcre/8.39/lib` in the compiler to find the correct version of PCRE. We are probably missing a more general way to decide where to look up libraries...

**Comment by jhass** [comment]
```
$ pkg-config --libs libpcre
-L/usr/local/Cellar/pcre/8.39/lib -lpcre
```

**Comment by asterite** [comment]
@jhass Ouch, I tried that but passing `pcre` instead... 😊 

Doing this works now:

```cr
@[Link(ldflags: "`pkg-config --libs libpcre`")]
lib LibPCRE
  # ...
end
```

I still wonder if there's a better way (I don't feel good putting that `pkg-config` there, though we also have it in a few other places, and more complex logic for openssl and llvm)

**Comment by jhass** [comment]
It'd be good to fallback to `-lpcre` if `pkg-config` is unavailable, though it's pretty standard to have.

Btw there's a `pkg-config --list-all`;)

**Comment by ysbaddaden** [comment]
Isn't the build command trying to run `pkg-config` by itself? The problem is maybe the `pcre` vs `libpcre`?

**Comment by RX14** [comment]
On my arch box, 361/1303 pkg-config entries begin with "lib". Seems like quite a non-standard.

**Comment by luislavena** [comment]
According to docs, `@[Link("libpcre")]` should perform `pkg-config` for you and determine the settings configuration. :confused:

**Comment by RX14** [comment]
@luislavena Wouldn't that would fall back to `-llibpcre` (which is incorrect) if pkg-config fails?

**Comment by luislavena** [comment]
BTW, `pkg-config pcre` returns error while `pkg-config libpcre` works, so perhaps the issue is that the package name differs from the linking library and the logic in `Program#pkg_config_flags` needs to be changed?

**Comment by ysbaddaden** [comment]
Yes. Maybe test `pkg-config name` then fallback on `pkg-config libname` then to `-lname`?

**Comment by RX14** [comment]
Looks like this should be a fairly easy low-hanging performance fruit in the stdlib?

**Comment by Blacksmoke16** [comment]
There was an issue with the benchmark in the first example for `pcre_jit_exec` case. I updated the result. It's still the slowest case.

After some additional research, we could possibly do what PHP did as they also use PCRE: https://github.com/php/php-src/blob/PHP-7.2/ext/pcre/php_pcre.c#L868-L876. Seems they have it enabled by default, except for `s390` OS (https://bugs.php.net/bug.php?id=73121), which we probably dont even (want to?) support. They then have some additional checks, versus using it for every context.

Another option would be to have it as part of a compiler flag, but if we do things right that's prob not needed. 

I'm thinking the best approach for now would be to just add this flag and call it a day. Can think about fine tuning things later, possibly after PCRE2 migration.
